sandbox/plugins/cryptodirectory/src/main/java/org/opensearch/index/store/CryptoDirectory.java

```diff
@@ -0,0 +1,530 @@
+     *
+     * @opensearch.internal
+     */
+    final class CryptoBufferedIndexInput extends BufferedIndexInput {
+        /** The maximum chunk size for reads of 16384 bytes. */
+        private static final int CHUNK_SIZE = 16384;
+        ByteBuffer tmpBuffer = ByteBuffer.allocate(CHUNK_SIZE);
+
+        /** the file channel we will read from */
+        protected final FileChannel channel;
+        /** is this instance a clone and hence does not own the file to close it */
+        boolean isClone = false;
+        /** start offset: non-zero in the slice case */
+        protected final long off;
+        /** end offset (start+length) */
+        protected final long end;
+        InputStream stream;
+        Cipher cipher;
+        final CryptoDirectory directory;
+
+        public CryptoBufferedIndexInput(String resourceDesc, FileChannel fc, IOContext context, Cipher cipher, CryptoDirectory directory)
+            throws IOException {
+            super(resourceDesc, context);
+            this.channel = fc;
+            this.off = 0L;
+            this.end = fc.size();
+            this.stream = Channels.newInputStream(channel);
+            this.cipher = cipher;
+            this.directory = directory;
+        }
+
+        public CryptoBufferedIndexInput(
+            String resourceDesc,
+            FileChannel fc,
+            long off,
+            long length,
+            int bufferSize,
+            Cipher old,
+            CryptoDirectory directory
+        ) throws IOException {
+            super(resourceDesc, bufferSize);
+            this.channel = fc;
+            this.off = off;
+            this.end = off + length;
+            this.isClone = true;
+            this.directory = directory;
+            this.stream = Channels.newInputStream(channel);
+            cipher = CipherFactory.getCipher(old.getProvider());
+            CipherFactory.initCipher(cipher, directory, Optional.of(old.getIV()), Cipher.DECRYPT_MODE, off);
+        }
+
+        /**
+         * {@inheritDoc}
+         */
+        @Override
+        public void close() throws IOException {
+            if (!isClone) {
+                stream.close();
+            }
+        }
+
+        /**
+         * {@inheritDoc}
+         */
+        @Override
+        public CryptoBufferedIndexInput clone() {
+            CryptoBufferedIndexInput clone = (CryptoBufferedIndexInput) super.clone();
+            clone.isClone = true;
+            clone.cipher = CipherFactory.getCipher(cipher.getProvider());
+            CipherFactory.initCipher(clone.cipher, directory, Optional.of(cipher.getIV()), Cipher.DECRYPT_MODE, getFilePointer() + off);
+            clone.tmpBuffer = ByteBuffer.allocate(CHUNK_SIZE);
+            return clone;
+        }
+
+        /**
+         * {@inheritDoc}
+         */
+        @Override
+        public IndexInput slice(String sliceDescription, long offset, long length) throws IOException {
```

Comment:
Is this slicing data to be decrypted?

Reply:
@stephen-crawford  yes. it returns an `IndexInput `which will return decrypted bytes when read.